File: src/core/execution_engine.py
import asyncio
import logging
import os
from typing import Any, Dict, Callable
from src.core.tool_tracker import ToolAttemptTracker

logger = logging.getLogger("src")


class ExecutionEngine:
    def __init__(
        self, assistant, tools, tool_parser, ui, instance_mode=False, tracker=None
    ):
        self.assistant = assistant
        self.tools = tools
        self.tool_parser = tool_parser
        self.ui = ui
        self.instance_mode = instance_mode
        self.tool_tracker = tracker or ToolAttemptTracker()
        self.turn_count = 0
        self.max_turns = 10000
        self.stop_requested = False

        self._detail_mappers: Dict[str, Callable[[Dict], str]] = {
            "readFile": lambda a: os.path.basename(a.get("path", "")),
            "writeFile": lambda a: os.path.basename(a.get("path", "")),
            "removeFile": lambda a: os.path.basename(a.get("path", "")),
            "replaceString": lambda a: os.path.basename(a.get("path", "")),
            "listDir": lambda a: a.get("path", ""),
            "create_dir": lambda a: a.get("path", ""),
            "renamePath": lambda a: f"{os.path.basename(a.get('origen', ''))} -> {os.path.basename(a.get('destino', ''))}",
            "pushMemory": lambda a: a.get("key", "unknown"),
            "pullMemory": lambda a: a.get("key", "unknown"),
            "run_instance": lambda a: a.get("target", "unknown"),
            "searchTextGlobal": lambda a: f"'{a.get('query', '')}'",
            "searchFilesPattern": lambda a: f"pattern: {a.get('pattern', '')}",
            "analyzePyright": lambda a: os.path.basename(a.get("path", "")),
            "runPython": lambda a: os.path.basename(a.get("script_path", "")),
        }

    async def execute_tool_call(self, name, args, call_data):
        if not self.tool_tracker.record_attempt(name, args):
            return (
                self.tool_parser.format_tool_response(
                    f"Tool loop detected: '{name}' called too many times.",
                    call_id=call_data.get("id"),
                    success=False,
                ),
                False,
            )

        if name not in self.tools:
            return (
                self.tool_parser.format_tool_response(
                    f"Unknown tool: {name}",
                    call_id=call_data.get("id"),
                    success=False,
                ),
                False,
            )

        if name == "run_instance" and self.instance_mode:
            return (
                self.tool_parser.format_tool_response(
                    "ERR: Recursion disabled.",
                    call_id=call_data.get("id"),
                    success=False,
                ),
                False,
            )

        detail = self._detail_mappers.get(name, lambda a: "")(args)
        self.ui.start_tool_monitoring(name, detail)
        await asyncio.sleep(0)

        try:
            tool_func = self.tools[name]
            if asyncio.iscoroutinefunction(tool_func):
                result = await tool_func(**args)
            else:
                result = await asyncio.to_thread(tool_func, **args)

            if name == "pullMemory":
                self.ui.display_message(
                    f"🧠 **Memento Pull**: Accessing node #{args.get('key')}"
                )
            elif name == "pushMemory":
                self.ui.display_message(f"✍️ **Memento Push**: {result}")
            elif name == "run_instance":
                self.ui.display_message(
                    f"**🚀 Instance Test Output**\n\n```text\n{result}\n```"
                )

            self.tool_tracker.record_success(name)
            self.ui.end_tool_monitoring(name, success=True, detail=detail)
            return (
                self.tool_parser.format_tool_response(
                    result, call_id=call_data.get("id"), success=True
                ),
                True,
            )
        except Exception as e:
            self.tool_tracker.record_failure(name)
            self.ui.end_tool_monitoring(name, success=False, detail=detail)
            return (
                self.tool_parser.format_tool_response(
                    f"ERR: {str(e)}", call_id=call_data.get("id"), success=False
                ),
                False,
            )

File: src_js/core/executionEngine.js
import path from 'path';
import { SmartToolParser } from './smartParser.js';
import { ToolAttemptTracker } from './toolTracker.js';

export class ExecutionEngine {
  constructor(assistant, tools, toolParser, ui, instanceMode = false, tracker = null) {
    this.assistant = assistant;
    this.tools = tools;
    this.toolParser = toolParser;
    this.ui = ui;
    this.instanceMode = instanceMode;
    this.toolTracker = tracker || new ToolAttemptTracker();
    this.turnCount = 0;
    this.maxTurns = 10000;
    this.stopRequested = false;

    this._detailMappers = {
      'readFile': (a) => path.basename(a.path || ''),
      'writeFile': (a) => path.basename(a.path || ''),
      'removeFile': (a) => path.basename(a.path || ''),
      'replaceString': (a) => path.basename(a.path || ''),
      'listDir': (a) => a.path || '',
      'createDir': (a) => a.path || '',
      'renamePath': (a) => path.basename(a.origin || '') + ' -> ' + path.basename(a.destination || ''),
      'pushMemory': (a) => a.key || 'unknown',
      'pullMemory': (a) => a.key || 'unknown',
      'runInstance': (a) => a.target || 'unknown',
      'searchTextGlobal': (a) => "'" + (a.query || '') + "'",
      'searchFilesPattern': (a) => 'pattern: ' + (a.pattern || ''),
      'analyzePyright': (a) => path.basename(a.path || ''),
      'runPython': (a) => path.basename(a.scriptPath || ''),
    };
  }

  async executeToolCall(name, args, callData) {
    if (!this.toolTracker.recordAttempt(name, args)) {
      return {
        response: this.toolParser.formatToolResponse(
          'Tool loop detected: ' + name + ' called too many times.',
          callData.id,
          false
        ),
        success: false,
      };
    }

    if (!(name in this.tools)) {
      return {
        response: this.toolParser.formatToolResponse(
          'Unknown tool: ' + name,
          callData.id,
          false
        ),
        success: false,
      };
    }

    if (name === 'runInstance' && this.instanceMode) {
      return {
        response: this.toolParser.formatToolResponse(
          'ERR: Recursion disabled.',
          callData.id,
          false
        ),
        success: false,
      };
    }

    // Safe detail mapping
    let detail = 'no details';
    try {
      const mapper = this._detailMappers[name] || (() => 'no details');
      detail = mapper(args);
    } catch (e) {
      detail = 'error mapping details';
    }
    
    this.ui.startToolMonitoring(name, detail);

    try {
      const toolFunc = this.tools[name];
      const result = await toolFunc(args);

      if (name === 'pullMemory') {
        this.ui.displayMessage('\ud83e\udde0 **Memento Pull**: Accessing node #' + args.key);
      } else if (name === 'pushMemory') {
        this.ui.displayMessage('\u270d\ufe0f **Memento Push**: ' + result);
      } else if (name === 'runInstance') {
        this.ui.displayMessage('\ud83d\ude80 **Instance Test Output**\n\n' + '```text\n' + result + '\n```');
      }

      this.toolTracker.recordSuccess(name);
      this.ui.endToolMonitoring(name, true, detail);

      return {
        response: this.toolParser.formatToolResponse(
          result,
          callData.id,
          true
        ),
        success: true,
      };
    } catch (e) {
      this.toolTracker.recordFailure(name);
      this.ui.endToolMonitoring(name, false, detail);

      return {
        response: this.toolParser.formatToolResponse(
          'ERR: ' + e.message,
          callData.id,
          false
        ),
        success: false,
      };
    }
  }
}